WooCommerce/Classes/Tools/BoldableTextParser.swift

```diff
@@ -0,0 +1,77 @@
+    ///
+    /// - Returns: A collection of boldable elements.
+    func parseBoldableElements(string: String) -> [BoldableElement] {
+        guard let regex = try? NSRegularExpression(pattern: boldTextPattern) else {
+            return []
+        }
+        let range = NSRange(string.startIndex..., in: string)
+
+        /// Finds all the ranges that match the regex for content within a bold region.
+        let matches: [NSTextCheckingResult] = regex.matches(in: string, options: [], range: range)
+
+        guard matches.isNotEmpty else {
+            return [.init(content: string, isBold: false)]
+        }
+
+        var elements: [BoldableElement] = []
+
+        var startingIndex = 0
+        matches.enumerated().forEach { index, match in
+            // Each `NSTextCheckingResult` has two ranges: the first one (index 0) is the range that includes the bold marks (**..**).
+            // The second one (index 1) is the range that excludes the bold marks, which corresponds to the bolded substring.
+            guard match.numberOfRanges == 2 else {
+                return
+            }
+
+            let matchingRange = match.range(at: 0)
+            let boldedContentRange = match.range(at: 1)
+
+            // Adds an element for the range from the last index and before the matching range, if non-empty.
+            let rangeBefore = startingIndex..<matchingRange.lowerBound
+            if rangeBefore.isEmpty == false {
+                let content = string[rangeBefore]
+                elements.append(.init(content: content, isBold: false))
+            }
+
+            // Adds an element for the matching range which should be bold.
+            let matchContent = string[boldedContentRange.lowerBound..<boldedContentRange.upperBound]
+            elements.append(.init(content: matchContent, isBold: true))
+
+            // Updates the last starting index so that the next range starts after the end of current matching range.
+            startingIndex = matchingRange.upperBound
+
+            // If the matching range is the last one, adds an element for the range after the last matching range, if non-empty.
+            let isLastMatchingRange = match == matches.last
+            if isLastMatchingRange {
+                let lastRange = matchingRange.upperBound..<string.count
+                if lastRange.isEmpty == false {
+                    let content = string[lastRange]
+                    elements.append(.init(content: content, isBold: false))
+                }
+            }
+        }
+        return elements
+    }
+}
+
+private extension String {
+    /// - Returns: A string subscript based on the given range.
+    subscript(range: Range<Int>) -> String {
+        let startIndex = index(self.startIndex, offsetBy: range.lowerBound)
+        let endIndex = index(self.startIndex, offsetBy: range.upperBound)
+        return String(self[startIndex..<endIndex])
+    }
+}
```

Comment:
> It looks like substring is deprecated in Swift 5, and the range has to be in the index form like what the String.subscript helper currently does (also suggested in this SO response) - do you have other suggestions on the replacement of String extension subscript(range: Range<Int>) -> String?

Ah, true, `substring` is deprecated. Because `NSTextCheckingResult` returns ranges as `NSRange` we can't directly use it with `String`, but we can convert it to swift `Range`. This will enable us to do all range calculations and later get a substring without index conversion.

Below is a suggestion that works without String extension.

```suggestion
        var startingIndex = string.startIndex
        matches.enumerated().forEach { index, match in
            // Each `NSTextCheckingResult` has two ranges: the first one (index 0) is the range that includes the bold marks (**..**).
            // The second one (index 1) is the range that excludes the bold marks, which corresponds to the bolded substring.
            guard match.numberOfRanges == 2 else {
                return
            }

            guard let matchingRange = Range(match.range(at: 0), in: string),
                  let boldedContentRange = Range(match.range(at: 1), in: string) else {
                return
            }

            // Adds an element for the range from the last index and before the matching range, if non-empty.
            let rangeBefore = startingIndex..<matchingRange.lowerBound
            if rangeBefore.isEmpty == false {
                let content = String(string[rangeBefore])
                elements.append(.init(content: content, isBold: false))
            }

            // Adds an element for the matching range which should be bold.
            let matchContent = String(string[boldedContentRange.lowerBound..<boldedContentRange.upperBound])
            elements.append(.init(content: matchContent, isBold: true))

            // Updates the last starting index so that the next range starts after the end of current matching range.
            startingIndex = matchingRange.upperBound

            // If the matching range is the last one, adds an element for the range after the last matching range, if non-empty.
            let isLastMatchingRange = match == matches.last
            if isLastMatchingRange {
                let lastRange = matchingRange.upperBound..<string.endIndex
                if lastRange.isEmpty == false {
                    let content = String(string[lastRange])
                    elements.append(.init(content: content, isBold: false))
                }
            }
        }
        return elements
    }
}
```

Reply:
thanks so much @eshurakov for your suggestions! I didn't have extensive experience with handling `String` and ranges and I really appreciate your insights 🙏 I applied your suggestions here https://github.com/woocommerce/woocommerce-ios/pull/3285/commits/da95e62728f7c19b2f138986110369a0d084b338!